src/librustc/middle/resolve_lifetime.rs

```diff
@@ -2066,27 +2070,78 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
         }
     }
 
+    fn lint_implicit_lifetimes_in_segment(&mut self,
+                                          segment_ident: ast::Ident,
+                                          generic_args: &'tcx hir::GenericArgs,
+                                          lifetime_refs: &[&'tcx hir::Lifetime]) {
+        let num_implicit_lifetimes = lifetime_refs.iter()
+            .filter(|lt| lt.name.is_implicit()).count();
+        if num_implicit_lifetimes == 0 {
+            return;
+        }
+
+        let mut err = self.tcx.struct_span_lint_node(
+            lint::builtin::ELIDED_LIFETIMES_IN_PATHS,
+            lifetime_refs[0].id, // FIXME: HirIdify #50928
+            segment_ident.span,
+            &format!("implicit lifetime parameters in types are deprecated"),
+        );
+
+        if num_implicit_lifetimes == 1 {
+            let (replace_span,
+                 suggestion) = if generic_args.args.len() == num_implicit_lifetimes &&
+                generic_args.bindings.is_empty() {
+                    // If there are no (non-implicit) generic args or bindings, our
+                    // suggestion includes the angle brackets
+                    (segment_ident.span.shrink_to_hi(), "<'_>")
+                } else {
+                    // Otherwise—sorry, this is kind of gross—we need to infer the
+                    // replacement point span from the generics that do exist
```

Comment:
I don't see a better way, but I get nervous around macros and things -- e.g., what will happen if somebody does:

`Ref<$x>`

in a macro? (I'm imagining that the span of `$x` might then come from some other place than the type itself...?) Maybe we want a test around that case? I think it'd be fine to disable suggestions there, but it'd be nice to ICE or suggest something mangled?